### PR TITLE
Add quasiquote for matchers and attributes

### LINKED ISSUE #17396: Better integration between MBE and procedural macros
For example, it should be easy to procedurally invoke the MBE engine's LHS matcher on an arbitrary token tree. This could take a lot of the tedium out of argument parsing in procedural macros.

### LINKED ISSUE #19674: Add quasiquote for attributes
[Sadness](https://github.com/servo/servo/blob/39271aa42d77ac7bbfabb1fd11a975265d645c99/components/plugins/jstraceable.rs#L17-L19)

### PR DESCRIPTION
**Pull request by pczarn**
Fixes #19674
Fixes #17396 (already closed, yeah)

cc @kmcallister , @cmr

### CODE REVIEW

File: src/libsyntax/ext/quote.rs [hunk 1]
```diff
@@ -683,27 +723,60 @@ fn mk_tt(cx: &ExtCtxt, tt: &ast::TokenTree) -> Vec<P<ast::Stmt>> {
...
+                                         mk_ast_path(cx, sp, "TtSequence"),
+                                         vec!(e_sp, e_rc_new));
+                let e_push =
+                    cx.expr_method_call(sp,
+                                        cx.expr_ident(sp, id_ext("tt")),
+                                        id_ext("push"),
+                                        vec!(e_tok));
+                vec!(cx.stmt_expr(e_push))
+            } else {
+                panic!("TtSequence in quote!")
```
Comment (kmcallister) [inline src/libsyntax/ext/quote.rs]: Could you move this to the top as

```rust
if !matcher {
    panic!("TtSequence in quote!");
}
```

and then un-indent the bulk of the code?

File: src/libsyntax/ext/quote.rs [hunk 2]
```diff
@@ -756,8 +833,23 @@ fn expand_tts(cx: &ExtCtxt, sp: Span, tts: &[ast::TokenTree])
 
     let stmt_let_tt = cx.stmt_let(sp, true, id_ext("tt"), cx.expr_vec_ng(sp));
 
-    let mut vector = vec!(stmt_let_sp, stmt_let_tt);
-    vector.extend(mk_tts(cx, &tts[]).into_iter());
+    vec!(stmt_let_sp, stmt_let_tt)
+}
+
+fn mk_tts(cx: &ExtCtxt, tts: &[ast::TokenTree], matcher: bool) -> Vec<P<ast::Stmt>> {
```
Comment (kmcallister) [inline src/libsyntax/ext/quote.rs]: Why does `mk_tts` return a vector of statements?

File: src/libsyntax/ext/quote.rs (hunk 1 above)
Comment (pczarn) [inline src/libsyntax/ext/quote.rs]: Done

File: src/libsyntax/ext/quote.rs (hunk 2 above)
Comment (pczarn) [inline src/libsyntax/ext/quote.rs]: `mk_tts` creates statements. It could return a FlatMap instead, but the signature would be complex.

Comment (pczarn) [inline src/libsyntax/ext/quote.rs]: Ah, the name is imprecise, I should change it

**Comment by kmcallister** [comment]
r=me with those two things

**Comment by pczarn** [comment]
Updated, tested.

**Comment by Manishearth** [comment]
(I might be forcebuilding the rollup, so the build may fail a couple of times here, sorry in advance)

**Comment by Manishearth** [comment]
In case the rollup causes a merge conflict, here is the fix: https://github.com/rust-lang/rust/commit/290caba2588ccd755f7b85a74d46f46d2ec75263

**Pull request merged**